File: src/main/cpp/2021/03/AoC2021_03.cpp
#include "assert.h"
#include <iostream>
#include <string>
#include <vector>
#include "../../aoc/aoc.hpp"
#include "../../aocd/aocd.hpp"

using namespace std;

const vector<string> TEST = {
"00100",
"11110",
"10110",
"10111",
"10101",
"01111",
"00111",
"11100",
"10000",
"11001",
"00010",
"01010"
};

class BitCount {
    private:
        uint ones;
        uint zeroes;

    public:
        BitCount(const uint ones, const uint zeroes) {
            this->ones = ones;
            this->zeroes = zeroes;
        }
        
        char leastCommon() const {
            return this->ones < this->zeroes ? '1' : '0';
        }

        char mostCommon() const {
            return this->ones >= this->zeroes ? '1' : '0';
        }
};

BitCount bitCounts(const vector<string> &strings, const uint pos) {
    int zeroes = 0;
    for (const string s : strings) {
        if (s[pos] == '0') {
            zeroes++;
        }
    }
    return BitCount(strings.size() - zeroes, zeroes);
}

uint ans(const string& s1, const string& s2) {
    return stoi(s1, 0, 2) * stoi(s2, 0, 2);
}

int part1(const vector<string> &input) {
    string gamma = "";
    string epsilon = "";
    for (uint i = 0; i < input[0].length(); i++) {
        BitCount bitCount = bitCounts(input, i);
        gamma += bitCount.mostCommon();
        epsilon += bitCount.leastCommon();
    }
    return ans(gamma, epsilon);
}

string reduce(vector<string> input, const bool mostCommon) {
    const uint size = input.size();
    for (uint pos = 0; pos < size; pos++) {
        const BitCount bitCount = bitCounts(input, pos);
        const char toKeep = mostCommon
            ? bitCount.mostCommon()
            : bitCount.leastCommon();
        vector<string> reduced;
        for (const string s : input) {
            if (s[pos] == toKeep) {
                reduced.push_back(s);
            }
        }
        if (reduced.size() == 1) {
            return reduced[0];
        }
        assert(!reduced.empty());
        input = move(reduced);
    }
    throw "illegal state";
}

int part2(vector<string> input) {
    const string o2 = reduce(input, true);
    const string co2 = reduce(input, false);
    return ans(o2, co2);
}

void samples() {
    assert(part1(TEST) == 198);
    assert(part2(TEST) == 230);
}

MAIN(2021, 3)

File: src/main/julia/AoC2021_03.jl
#! /usr/bin/env julia
#

include("aoc_main.jl")
if abspath(PROGRAM_FILE) == @__FILE__
    include("aocd.jl")
    using .Aocd
end

module AoC2021_03

include("aoc.jl")

function bitcount(input, pos)
    size = length(input)
    c = count(==('1'), [input[r][pos] for r = 1:size])
    return c >= size / 2 ? ['1', '0'] : ['0', '1']
end

function ans(value1, value2)
    return prod(parse(Int, value, base=2)
                for value ∈ [value1, value2])
end

function part1(input)
    bc = [bitcount(input, pos) for pos ∈ 1:sizeof(input[1])]
    γ, ϵ = [join([x[i] for x ∈ bc]) for i ∈ 1:2]
    return ans(γ, ϵ)
end

function part2(input)
    function reduce(input, i)
        strings = copy(input)
        pos = 1
        while length(strings) > 1
            keep = bitcount(strings, pos)[i]
            strings = strings[findall(s -> s[pos] == keep, strings)]
            pos += 1
        end
        return strings[1]
    end
    o2, co2 = [reduce(input, i) for i ∈ 1:2]
    return ans(o2, co2)
end

TEST = @aoc_splitlines("""
00100
11110
10110
10111
10101
01111
00111
11100
10000
11001
00010
01010
""")

function samples()
    @assert part1(TEST) == 198
    @assert part2(TEST) == 230
end

end # module AoC2021_03

aoc_main(@__FILE__, ARGS, 2021, 3)
